File: packages/protocol/src/hello.ts
/** The `hello` handshake: what the binary speaks and can do. */

import { codes, ProtocolError } from './error.ts';
import { isWellFormedUnicode } from './json-token.ts';
import { arrayValues, assertClosedObject, ownDataValue } from './shape.ts';

/** Capability advertised when `workflow.signal.submit` is available. */
export const CAPABILITY_WORKFLOW_SIGNAL_SUBMIT = 'workflow.signal.submit';
/** Capability advertised when `workflow.signal.reconcile` is available. */
export const CAPABILITY_WORKFLOW_SIGNAL_RECONCILE = 'workflow.signal.reconcile';

/** Informational identity of the responding package. */
export interface PackageInfo {
  readonly name: string;
  readonly version: string;
}

/**
 * The `hello` response payload. Adapters decide compatibility from
 * `protocolVersion` and `capabilities`, never from `package.version`.
 */
export interface HelloInfo {
  readonly protocolVersion: number;
  readonly journalSchemaVersion: number;
  readonly capabilities: readonly string[];
  readonly package: PackageInfo;
}

const U32_MAX = 4294967295;
const CAPABILITY = /^[a-z][a-z0-9]*(\.[a-z][a-z0-9]*)*$/;

function isVersion(value: unknown): value is number {
  return typeof value === 'number' && Number.isInteger(value) && value >= 1 && value <= U32_MAX;
}

function isCapability(value: unknown): value is string {
  return typeof value === 'string' && value.length <= 128 && CAPABILITY.test(value);
}

/** Package-internal validator and fresh-wire builder shared by decode and encode. */
export function buildHelloInfo(payload: unknown): HelloInfo {
  const fail = (message: string) => new ProtocolError(codes.INVALID_PAYLOAD, message);
  const allowed = ['protocolVersion', 'journalSchemaVersion', 'capabilities', 'package'];
  assertClosedObject(payload, allowed, fail, 'hello payload');
  const protocolVersion = ownDataValue(payload, 'protocolVersion', fail, 'hello payload');
  const journalSchemaVersion = ownDataValue(payload, 'journalSchemaVersion', fail, 'hello payload');
  const capabilitiesValue = ownDataValue(payload, 'capabilities', fail, 'hello payload');
  const pkg = ownDataValue(payload, 'package', fail, 'hello payload');
  if (!isVersion(protocolVersion)) {
    throw fail('protocolVersion must be an integer between 1 and 4294967295');
  }
  if (!isVersion(journalSchemaVersion)) {
    throw fail('journalSchemaVersion must be an integer between 1 and 4294967295');
  }
  const capabilities = arrayValues(capabilitiesValue, fail, 'capabilities');
  for (let index = 0; index < capabilities.length; index += 1) {
    if (!isCapability(capabilities[index])) {
      throw fail(
        'capabilities must be lowercase dot-separated names (^[a-z][a-z0-9]*(\\.[a-z][a-z0-9]*)*$, at most 128 bytes)',
      );
    }
    for (let earlier = 0; earlier < index; earlier += 1) {
      if (capabilities[earlier] === capabilities[index]) {
        throw fail('capabilities must not repeat');
      }
    }
  }
  assertClosedObject(pkg, ['name', 'version'], fail, 'package');
  const name = ownDataValue(pkg, 'name', fail, 'package');
  const version = ownDataValue(pkg, 'version', fail, 'package');
  if (
    typeof name !== 'string' ||
    typeof version !== 'string' ||
    !isWellFormedUnicode(name) ||
    !isWellFormedUnicode(version)
  ) {
    throw fail('package.name and package.version must be strings');
  }
  return {
    protocolVersion,
    journalSchemaVersion,
    capabilities: capabilities as string[],
    package: { name, version },
  };
}

/** Decodes a `hello` payload, rejecting anything outside the closed schema. */
export function decodeHelloInfo(payload: unknown): HelloInfo {
  return buildHelloInfo(payload);
}

/** Why a binary is not usable by this adapter. */
export interface Incompatibility {
  readonly reason: 'protocol_version' | 'missing_capability';
  readonly detail: string;
}

/**
 * Compatibility is decided by protocol version and capabilities only
 * (ADR-0003, ADR-0008). Returns `undefined` when compatible.
 */
export function checkCompatibility(
  hello: HelloInfo,
  required: { readonly protocolVersion: number; readonly capabilities: readonly string[] },
): Incompatibility | undefined {
  if (hello.protocolVersion !== required.protocolVersion) {
    return {
      reason: 'protocol_version',
      detail: `binary speaks protocol ${hello.protocolVersion}; this adapter requires ${required.protocolVersion}`,
    };
  }
  for (let requiredIndex = 0; requiredIndex < required.capabilities.length; requiredIndex += 1) {
    const capability = required.capabilities[requiredIndex];
    let present = false;
    for (let actualIndex = 0; actualIndex < hello.capabilities.length; actualIndex += 1) {
      if (hello.capabilities[actualIndex] === capability) {
        present = true;
        break;
      }
    }
    if (!present) {
      return { reason: 'missing_capability', detail: `binary lacks capability \`${capability}\`` };
    }
  }
  return undefined;
}

File: crates/aizu-protocol/src/hello.rs
//! The `hello` handshake: what this binary speaks and can do.

use serde::{Deserialize, Serialize};

/// Capability advertised when `workflow.signal.submit` is available.
pub const CAPABILITY_WORKFLOW_SIGNAL_SUBMIT: &str = "workflow.signal.submit";

/// The `hello` response payload. Adapters decide compatibility from
/// `protocol_version` and `capabilities`, never from `package.version`
/// (ADR-0003, ADR-0008).
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct HelloInfo {
    /// Wire protocol version this binary speaks.
    pub protocol_version: u32,
    /// Journal schema version this binary reads and writes.
    pub journal_schema_version: u32,
    /// Request kinds this binary accepts beyond `hello`.
    pub capabilities: Vec<String>,
    /// Informational package identity; not used for compatibility.
    pub package: PackageInfo,
}

impl HelloInfo {
    /// Contract checks past serde's shape (kept identical to
    /// `spec/protocol/v1/schemas/hello.response.schema.json`): versions start
    /// at 1, capability names are well formed and unique. The capability
    /// *list* stays open so a client can still decode the handshake of a
    /// newer binary; compatibility is decided afterwards by the caller.
    pub fn validate(&self) -> Result<(), String> {
        if self.protocol_version == 0 {
            return Err("protocolVersion must be at least 1".to_owned());
        }
        if self.journal_schema_version == 0 {
            return Err("journalSchemaVersion must be at least 1".to_owned());
        }
        let mut seen = std::collections::BTreeSet::new();
        for capability in &self.capabilities {
            if !valid_capability(capability) {
                return Err(format!(
                    "capability `{capability}` must match ^[a-z][a-z0-9]*(\\.[a-z][a-z0-9]*)*$ (at most 128 bytes)"
                ));
            }
            if !seen.insert(capability.as_str()) {
                return Err(format!("capability `{capability}` is listed twice"));
            }
        }
        Ok(())
    }
}

/// Lowercase dot-separated tokens, at most 128 bytes.
fn valid_capability(name: &str) -> bool {
    name.len() <= 128
        && !name.is_empty()
        && name.split('.').all(|token| {
            let mut chars = token.chars();
            chars.next().is_some_and(|first| first.is_ascii_lowercase())
                && chars.all(|c| c.is_ascii_lowercase() || c.is_ascii_digit())
        })
}

/// Informational identity of the responding package.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct PackageInfo {
    /// Package name, e.g. `aizu`.
    pub name: String,
    /// Package version, e.g. `0.1.0`.
    pub version: String,
}

#[cfg(test)]
mod tests {
    use super::*;

    fn hello(capabilities: &[&str]) -> HelloInfo {
        HelloInfo {
            protocol_version: 1,
            journal_schema_version: 1,
            capabilities: capabilities.iter().map(ToString::to_string).collect(),
            package: PackageInfo {
                name: "aizu".to_owned(),
                version: "0.1.0".to_owned(),
            },
        }
    }

    #[test]
    fn versions_start_at_one_and_capabilities_are_wellformed_and_unique() {
        assert!(
            hello(&[CAPABILITY_WORKFLOW_SIGNAL_SUBMIT])
                .validate()
                .is_ok()
        );
        // A newer binary may advertise more capabilities; decoding stays open.
        assert!(
            hello(&[CAPABILITY_WORKFLOW_SIGNAL_SUBMIT, "workflow.status.read"])
                .validate()
                .is_ok()
        );

        let mut zero = hello(&[]);
        zero.protocol_version = 0;
        assert!(zero.validate().is_err());
        let mut zero = hello(&[]);
        zero.journal_schema_version = 0;
        assert!(zero.validate().is_err());

        for bad in ["Workflow.Signal", "workflow..submit", "", "1abc", "a b"] {
            assert!(hello(&[bad]).validate().is_err(), "{bad:?}");
        }
        let long = format!("w{}", "a".repeat(128));
        assert!(hello(&[&long]).validate().is_err());
        assert!(
            hello(&[
                CAPABILITY_WORKFLOW_SIGNAL_SUBMIT,
                CAPABILITY_WORKFLOW_SIGNAL_SUBMIT
            ])
            .validate()
            .is_err()
        );
    }
}
